src/box/box.h

```diff
@@ -788,6 +788,22 @@ box_iproto_override(uint32_t req_type, iproto_handler_t handler,
 API_EXPORT int64_t
 box_info_lsn(void);
 
+typedef struct slab_info
```

Comment:
Some kind of extensibility needs to be considered here.
Probably the `box.slab.info()` output will be changed/extended in scope of:
- https://github.com/tarantool/tarantool/issues/10274#issuecomment-2242827465

So the C API should be able to support the new output without breaking the binary compatibility.

Reply:
@Totktonada WDYT about the suggested C API? Maybe, we should make the new structure extendable (reserve some padding for future extensions or allocate it on the region)? Another alternative is to introduce a function that would return the value of a particular memory usage counter, for example:

```C
enum box_slab_info_type {
    BOX_SLAB_INFO_ITEMS = 0,
    BOX_SLAB_INFO_INDEX = 1,
    /* ... */
};

uint64_t
box_slab_info(enum box_slab_info_type type);
```

Reply:
We definitely need this API be extendable. Padding is a valid option, but I think it is not the best option for informational API -- we may want to add many values to it.

The function that returns a specific counter looks better. However, I'm not a big fan of `enum`'s in the public API for such purposes. It obligates a user to use the new header version to have ability to use new metrics. Or define this enum on its side.

I think that string IDs fit better here. `strcmp()` on a few-bytes-long string is not a heavy thing. But it allows to compile against an old header version and get new metrics if actual tarantool version in runtime is the new one.

If this API is designing for logging/metrics reporting (or other way to give them to a human), it may be useful to get all the metrics disregarding whether they're known on the user side. We can return an array of key-value pairs `{.name = "items_size", .value = 42}`. This way getting many values at once doesn't require many calls. We can add two functions: for integer values and for double ones. Or return a type in the array:

```c
struct box_slab_info_item {
	const char *name;
	const char *type;
	union {
		uint64_t u64;
		double d;
	};
};

/*
 * Returns an array of metrics:
 *
 * [
 *     {"items_size", "uint64_t", 42},
 *     {"items_used", "uint64_t", 42},
 *     {"items_ratio", "double", 0.5},
 *     <...>
 * ]
 *
 * The array size is written to @a size.
 *
 * Allocated on box region, see box_region_used() and
 * box_region_truncate().
 */
struct box_slab_info_item *
box_slab_info(size_t *size);
```

Reply:
> I think that string IDs fit better here. `strcmp()` on a few-bytes-long string is not a heavy thing.

The problem with this approach is that fetching a particular counter value is inefficient and inconvenient because it requires scanning of all returned items. AFAIU they're planning to use this function to get the actual memory usage on every batch insertion.

Reply:
The option with enum suits us better. We need the fastest possible API.